packages/lesswrong/lib/ultraFeed/ultraFeedThreadHelpers.ts

```diff
@@ -19,504 +18,420 @@ import {
-    return [];
-  }
-
-  return buildCommentThreads(topLevelId);
-}
-
-/**
- * Builds comment threads for the UltraFeed using raw comment data
+ * Builds comment threads for the UltraFeed using the current algorithm.
+ * Fetches data, scores comments, builds threads, ranks threads, prepares for display.
  */
 export async function getUltraFeedCommentThreads(
   context: ResolverContext,
   limit = 20,
-): Promise<FeedCommentsThread[]> {
+  settings: UltraFeedSettingsType,
+  servedThreadHashes: Set<string> = new Set()
+): Promise<PreparedFeedCommentsThread[]> { // Return array of new type
   const userId = context.userId;
-
   if (!userId) {
-    // eslint-disable-next-line no-console
-    console.warn("getUltraFeedCommentThreads: No user ID provided. Returning empty array.");
     return [];
   }
 
-  const commentsRepo = context.repos.comments
+  // --- Step 1: Fetch Data ---
+  const commentsRepo = context.repos.comments;
+  const rawCommentsData = await commentsRepo.getCommentsForFeed(userId, 1000);
 
-  const candidates = await commentsRepo.getCommentsForFeed(userId, 500);
-  const threads = getAllCommentThreads(candidates);
-  const prioritizedThreadInfos = prioritizeThreads(threads);
+  // --- Step 2: Score Individual Comments ---
+  const relevantSettings = { 
+    commentDecayFactor: settings.commentDecayFactor, 
+    commentDecayBiasHours: settings.commentDecayBiasHours, 
+    commentSeenPenalty: settings.commentSeenPenalty,
+    quickTakeBoost: settings.quickTakeBoost 
+  };
+  const scoredComments = scoreComments(rawCommentsData, relevantSettings); 
 
-  const displayThreads = prioritizedThreadInfos
-    .slice(0, limit * 2)
-    .map(info => prepareCommentThreadForResolver(info))
-    .filter(thread => thread.comments.length > 0);
+  // --- Steps 3 & 4a: Build and Score Threads --- 
+  const threadScoringSettings = {
+    threadScoreAggregation: settings.threadScoreAggregation,
+    threadScoreFirstN: settings.threadScoreFirstN
+  };
+  const allScoredThreads = buildAndScoreThreads(scoredComments, threadScoringSettings); 
+
+  // --- Step 4b: Select Best Threads --- 
+  const finalRankedThreads = selectBestThreads(allScoredThreads); 
+
+  // --- Step 5: Prepare for Display --- 
+  // Filter out served threads
+  const unservedRankedThreads = finalRankedThreads.filter(rankedThreadInfo => {
+    const thread = rankedThreadInfo.thread;
+    if (!thread || thread.length === 0) return false;
+    const commentIds = thread.map(c => c.commentId);
+    const threadHash = UltraFeedEventsRepo.generateThreadHash(commentIds); // Use static method
+    return !servedThreadHashes.has(threadHash);
+  });
+  
+  const displayThreads = unservedRankedThreads
+    .slice(0, limit) 
+    .map(rankedThreadInfo => prepareThreadForDisplay(rankedThreadInfo))
+    .filter(Boolean) as PreparedFeedCommentsThread[];
```

Comment:
Why does this need a cast?

Reply:
"in case the typechecker isn't good enough" is what Shoggoth told me when I asked the same thing

Reply:
idgi